app/main.py

```diff
@@ -0,0 +1,380 @@
+    db.commit()
+    return {"completed": db_step}
+
+
+@app.get("/pipelines/", response_class=HTMLResponse)
+async def list_pipelines(request: Request, db: Session = Depends(get_db)):
+    """Controller for rendering the list of pipelines."""
+    pipelines = crud.get_pipelines(db)
+    pipeline_types = list(config["pipelines"].keys())
+    context = {"request": request,
+               "pipelines": pipelines,
+               "pipeline_types": pipeline_types}
+    return templates.TemplateResponse("list_pipelines.html", context)
+
+
+@app.get("/pipeline/{pipeline_id}", response_class=HTMLResponse)
+async def query_pipeline_by_id(pipeline_id: int,
+                               request: Request,
+                               db: Session = Depends(get_db)
+                               ):
+    """Controller for rendering a single pipeline."""
+    db_pipeline = crud.get_pipeline_by_id(db, pipeline_id)
+    if not db_pipeline:
+        raise HTTPException(
+            status_code=404,
+            detail=f"Pipeline with {pipeline_id=} does not exist."
+        )
+    context = {"request": request, "pipeline": db_pipeline}
+    return templates.TemplateResponse("view_pipeline.html", context)
+
+
+@app.get("/create_pipeline_form/{pipeline_type}", response_class=HTMLResponse)
+async def create_pipeline_form(pipeline_type: str, request: Request):
+    """Controller to render the create new pipeline form of a given type."""
+    params = config["pipelines"][pipeline_type]["params"]
+    context = {"request": request,
+               "params": params,
+               "pipeline_type": pipeline_type
+               }
+    return templates.TemplateResponse("new_pipeline.html", context)
+
+
+@app.post("/create_pipeline/", response_class=HTMLResponse)
+async def add_pipeline_from_form(request: Request,
+                                 db: Session = Depends(get_db),
+                                 ):
+    """Controller to process the response from create new pipeline form."""
+    form_data = await request.form()
+    form_data = jsonable_encoder(form_data)
+    prereq_pipe = form_data.pop("prereq_pipe")
+    prereq_step = form_data.pop("prereq_step")
+    pipeline = schemas.PipelineCreate(
+        template=form_data.pop("template"),
+        comment=form_data.pop("comment"),
+        prereq_pipe=int(prereq_pipe) if prereq_pipe != "" else None,
+        prereq_step=int(prereq_step) if prereq_step != "" else None,
+        params=form_data,
+    )
+    db_pipeline = crud.create_pipeline(db, pipeline)
+    context = {"request": request, "pipeline": db_pipeline}
+    return templates.TemplateResponse("view_pipeline.html", context)
+
+
+@app.get("/edit_pipeline/{pipeline_id}", response_class=HTMLResponse)
+async def edit_pipeline(pipeline_id: int,
+                        request: Request,
+                        db: Session = Depends(get_db)
+                        ):
+    """Controller to render the edit pipeline form."""

```

Comment:
"Controller" as in MVC? Wouldn't that make this a "view"?

Reply:
Nope, this is the "controller" in MVC, because it controls the flow of the app and of the data. The "view" part is the templates/edit_pipeline.html, while the "model" part is what happens in the crud.py, models.py and schemas.py.